### ISSUE TITLE
Account for ISUPPORT information in irc_MODE when parsing mode strings

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3296]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Wolf reported|
|-|-|
|Trac ID|trac#3296|
|Type|enhancement|
|Created|2008-06-14 15:09:49Z|
|Branch|https://github.com/twisted/twisted/tree/isupport-aware-mode-parsing-3296-2|

[#3230](https://github.com/twisted/twisted/issues/3230) introduced an "hardcoded" dict with some common modes as keys and a tuple of 2 bools as value:

```
_modeAcceptsArg = { 
  'o': (True, True),    # op/deop a user 
  'v': (True, True),    # voice/devoice a user 
  'b': (True, True),    # ban/unban a user/mask 
  'l': (True, False),   # set the user limit to channel 
  'k': (True, False),   # set a channel key (password) 
  't': (False, False),  # only ops set topic 
  's': (False, False),  # secret channel 
  ...
}
```

This dict is used by the irc_MODE method to know if a mode accept an argument or not.
The ISUPPORT message (see [#3285](https://github.com/twisted/twisted/issues/3285)) send a list with the current supported modes divided in 4 comma-separated groups, where the modes in the first 2 groups always accept an arg (when set and removed) the 3d group only when the mode is set and the 4th never accept args.

I suggest to use directly this list or replace the current dict with a list of 4 strings with the modes (if the server sends `CHANMODES=b,k,l,cd` the list will be `['b', 'k', 'l', 'cd']`). If we use a list we should save it as an instance attribute. The irc_MODE method will able to use a simple function like

```
def _modeAcceptArg(mode, set=True):
  # if we use the CHANMODES value sent by isupport:
  if mode not in self.server_capab['CHANMODES']: return False
  modes = self.server_capab['CHANMODES'].split(',')
  # otherwise, if we use a list
  if mode not in ''.join(self.chanmodes): return False
  modes = self.chanmodes
  # --
  if set and mode in modes[0] or mode in modes[1] or mode in modes[2]:
    return True
  else:
    return False
  if not set and mode in modes[0] or mode in modes[1]:
    return True
  else:
    return False
```

A default list of modes can either added as another instance attributes or saved inside the function (like `if 'CHANMODES' not in self.server_capab: modes = 'b,k,l,c'`).
So the modes supported by the server can be:
  1. the comma-separated string saved in server_cabap['CHANMODES'] (server_cabap is the dict that contains all the values of ISUPPORT)
  2. the `chanmodes` instance attribute as a list of 4 strings, saved when the ISUPPORT message is received

We also need a default list/string of modes that can be:
  1. saved as a class attribute
  2. hardcoded in the function (like `if 'CHANMODES' not in self.server_capab: modes = 'b,k,l,c'`).

Probably the best combination is 1.-1.

Attachments:

* [isupport-aware-mode-parsing.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7a9/7a98fad963cc52c4b78d76d75ee8ca0bc376a418/b2b1636c71d56cd46309487381ea6c78279f6922.diff) (16679 bytes) - added by jonathanj on 2009-05-21 18:32:27Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3296 3296
type__enhancement enhancement
reporter__Wolf Wolf
priority__normal normal
milestone__ 
branch__branches_isupport_aware_mode_parsing_3296_2 branches/isupport-aware-mode-parsing-3296-2
branch_author__jonathanj__Wolf jonathanj, Wolf
status__closed closed
resolution__fixed fixed
component__words words
keywords__irc_IRCClient_irc_MODE_modeChanged irc IRCClient irc_MODE modeChanged
time__1213456189000000 1213456189000000
changetime__1260457546000000 1260457546000000
version__None None
owner__jonathanj jonathanj
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Wolf commented|
|-|-|

Probably is also better to move all the 15 tests in a specific class instead of having them in the ModeTestCase class (also I'm not sure why is called ModeTestCase, it actually tests privmsg, MOTD, bounce, ISupport and other things)

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Wolf commented|
|-|-|

A comma-separated list of default modes can be saved in `server_capab['CHANMODES']` and replaced if and when the server sends the ISUPPORT message. The _modeSupportArg function can use the information saved in `server_capab['CHANMODES']` and in `server_capab['PREFIX']` to know if the mode accepts an argument. Doing this we avoid to have other instance attributes and duplicated informations (the comma-separated value in server_capab and the list as an attribute) but the function will have to parse the data every time (not a big problem though) and the user won't be able to know if `server_capab['CHANMODES']` contains the default values or the real ones sent by the server (unless if he checks if the ISUPPORT message has the CHANMODES parameter when it is received).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj removed owner|
|-|-|

I've attached a patch, whose changes mostly include:

  * Add a free-standing function to parse the parameters of an IRC mode message, optionally specifying modes that require parameters. The `irc_MODE` handler uses any information gleaned from ISUPPORT to parse the message parameters.

  * Most tests in `ModeTestSuite` are nonsensical, they use invented modes that change whether they take parameters or not from test to test. Also, half of `ModeTestSuite` seems to be for testing other server messages (not related to modes.)

  * It would seem, according to `ModeTestSuite` (whose actual value is somewhat dubious at the moment), that the length of the `args` param (to `modeChanged`) is not to match the number of modes. This potentially means that if a non-argument-taking and an argument-taking mode are passed (in that order), it would be a lot more work to figure out which argument belongs to which mode. A better approach would be to include arguments for every mode, using None for modes that take no argument (as `parseModes` does), but it would probably break backwards compatibility to change the way `modeChanged` is called.

This patch depends on [#3285](https://github.com/twisted/twisted/issues/3285) for ISUPPORT parameters.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_6"></a>|@jonathanj set owner to @jonathanj|
|-|-|

I'm taking this out of review until [#3285](https://github.com/twisted/twisted/issues/3285) has been merged, since the API has changed.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_7"></a>|@jonathanj commented|
|-|-|

(In [27496]) Branching to 'isupport-aware-mode-parsing-3296'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_8"></a>|@jonathanj commented|
|-|-|

(In [27498]) Remove _modeAcceptsArg. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_9"></a>|@jonathanj removed owner|
|-|-|

There is at one intentional XXX comment in the tests, that I'd like some input on. Support for `_modeAcceptsArg` has been dropped, I don't think it can make any sense given the way it merged user and channel modes together. Perhaps there is still room for some kind of "and here are some additional modes with arguments" mechanism, but I think that ISUPPORT is widely supported enough not to really warrant the extra complexity.

Most of my motivation is a few comments up, in case there are some things that are not clear.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. `ModeParsingTests` needs a class docstring.  If it talks about `irc.parseModes`, I probably won't make you rewrite all of that class's method docstrings in the active voice.
  1. But, ugg, all of the non-exception test docstrings are weird.  They're a little informative, since they at least explain how a human might think about the mode string being tested, but they should talk about the expected results as well.
  1. `assertEquals` is preferred over `assertEqual`
  1. In the implementation of `parseModes`, use tuples as the right-hand argument to string interpolation so it won't fail when someone gives you particularly bogus values.
  1. The return docs for `parseModes` are a bit ugly.  I'd drop the `@rtype` and just describe what's being returned with english.
  1. For the first condition in `irc_MODE`... What if it's another user's modes being changed?
  1. As far as the `XXX` goes... Be lenient in what you accept, etc.  I can't think of what other interpretation there might be for this error, so it seems okay.  Does some real software produce this?  Or was it just an obvious thing to test based on how the implementation went?
  1. The docstring for `getUserModeParams` says _This is only channel status modes._ which I find confusing.
  1. As far as dropping `_modeAcceptsArgs` goes... The main issue I might have here is that our very own IRC server does not emit _ISUPPORT_ lines.  It also doesn't do a whole lot with modes, so maybe this doesn't matter.  Still, the RFC defines some modes, and defines which take arguments.  Couldn't we default to these settings if no _ISUPPORT_ is received?  Would it make sense for `ServerSupportedFeatures` to include defaults for `'CHANMODES'`?  It already has a `'PREFIX'` default, after all.
 
Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj commented|
|-|-|

(In [27524]) Improve docstrings and use assertEquals. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_12"></a>|@jonathanj set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_10):
>   1. `ModeParsingTests` needs a class docstring.  If it talks about `irc.parseModes`, I probably won't make you rewrite all of that class's method docstrings in the active voice.
>   2. But, ugg, all of the non-exception test docstrings are weird.  They're a little informative, since they at least explain how a human might think about the mode string being tested, but they should talk about the expected results as well.

I've made some improvements to the method docstrings. However, I'm not sure what else to add for the class docstring or how to improve some of the less stellar method docstrings, everything seems so awkward. Some help here would be appreciated.

>   3. `assertEquals` is preferred over `assertEqual`
>   4. In the implementation of `parseModes`, use tuples as the right-hand argument to string interpolation so it won't fail when someone gives you particularly bogus values.
>   5. The return docs for `parseModes` are a bit ugly.  I'd drop the `@rtype` and just describe what's being returned with english.

Done.

>   7. As far as the `XXX` goes... Be lenient in what you accept, etc.  I can't think of what other interpretation there might be for this error, so it seems okay.  Does some real software produce this?  Or was it just an obvious thing to test based on how the implementation went?

I think it's okay to be lenient, it just seemed like something to test based on how obscure it is. I would hope that IRC servers normalize mode strings.

>   6. For the first condition in `irc_MODE`... What if it's another user's modes being changed?
>   8. The docstring for `getUserModeParams` says _This is only channel status modes._ which I find confusing.

I think we were both confused here. For the first condition in `irc_MODE`, that means we're setting modes on our own user (+i for invisible, +x for hostmask cloaking, etc.) which is different from setting a channel mode that affects our user (+o for channel operator, etc.). `getUserModeParams` should be returning UMODE (modes we set on our own user) modes that accept parameters, as far as I know there are none of these, there are certainly none specified in the _ISUPPORT_ draft and all the ones in RFC1459 take no parameters.

So I've changed `getUserModeParams` to return no modes, someone can override this if they need it, and `getChannelModeParams` to include what `getUserModeParams` used to do.

>   9. As far as dropping `_modeAcceptsArgs` goes... The main issue I might have here is that our very own IRC server does not emit _ISUPPORT_ lines.  It also doesn't do a whole lot with modes, so maybe this doesn't matter.  Still, the RFC defines some modes, and defines which take arguments.  Couldn't we default to these settings if no _ISUPPORT_ is received?  Would it make sense for `ServerSupportedFeatures` to include defaults for `'CHANMODES'`?  It already has a `'PREFIX'` default, after all.

The lack of _ISUPPORT_ is both unfortunate and fortunate. Fortunate because this seems like it could have been a catastrophe to omit, unfortunate in that I have to do extra work. ;-)

The _ISUPPORT_ draft explicitly mentions that there is no default for `CHANMODES`, which personally doesn't make a lot of sense since RFC1459 obviously specifies some defaults.

After having written and re-written and edited and re-edited a huge piece of text discussing ideas for dealing with this, I think it would be best to rather have this conversation on IRC.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. I think `ModeParsingTests` is missing a case, eg `parseModes('+ksbb', ['key', 'nick', 'other'])`.  From the docstring, it sounds like this is what `test_parameters` is going to cover, but then it doesn't.
  1. Regarding the call to `getUserModeParams` in `irc_MODE` - I think I understand now: users aren't told about mode changes to other users, only about mode changes to themselves and to channels?  A comment might help clarify that.
  1. If I comment out the `PREFIX`-related parts of `getChannelModeParams`, none of the test fail.  Some more tests for this are probably in order.
  1. On IRC we talked about retaining support for the things `_modeAcceptsArgs` is supporting and about how we don't have to actually keep spporting `_modeAcceptsArgs` itself for this if we don't want to (and we probably don't).  The way the RFC goes out of its way to state there is no default value for `CHANMODES` is annoying and makes me wonder if my suggestion above is really the right thing to do.  The alternative seems to be to continue to maintain these defaults separately, which would just be confusing I thing.  So maybe giving `CHANMODES` a default is the right thing.  At least add a comment about why we're going against the rfc (is there actually an rfc for isupport, or only a draft?)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_14"></a>|@jonathanj commented|
|-|-|

(In [27535]) Default CHANMODES to account for the stuff that _modeAcceptsArg used to do. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_15"></a>|@jonathanj commented|
|-|-|

(In [27536]) Improve test coverage. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_16"></a>|@jonathanj set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_13):
>   1. I think `ModeParsingTests` is missing a case, eg `parseModes('+ksbb', ['key', 'nick', 'other'])`.  From the docstring, it sounds like this is what `test_parameters` is going to cover, but then it doesn't.

Technically `-klbb` is a testcase for this, since `k` and `l` don't take arguments when being removed, but I've added a test for mixing a non-argument setting mode amongst argument-taking setting modes.

>   2. Regarding the call to `getUserModeParams` in `irc_MODE` - I think I understand now: users aren't told about mode changes to other users, only about mode changes to themselves and to channels?  A comment might help clarify that.

Added.

>   3. If I comment out the `PREFIX`-related parts of `getChannelModeParams`, none of the test fail.  Some more tests for this are probably in order.

I've added a test in `ClientImplementationTests`.

>   4. On IRC we talked about retaining support for the things `_modeAcceptsArgs` is supporting and about how we don't have to actually keep spporting `_modeAcceptsArgs` itself for this if we don't want to (and we probably don't).  The way the RFC goes out of its way to state there is no default value for `CHANMODES` is annoying and makes me wonder if my suggestion above is really the right thing to do.  The alternative seems to be to continue to maintain these defaults separately, which would just be confusing I thing.  So maybe giving `CHANMODES` a default is the right thing.  At least add a comment about why we're going against the rfc (is there actually an rfc for isupport, or only a draft?)

Done. One thing occured to me: `_modeAcceptsArg` has a key for the `h` channel mode (half-ops) which is not specified by any RFC. I'm wondering if this should be added to the `PREFIX` default (which would go against the ISUPPORT draft) in order to maximise backward compatibility.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_17"></a>|@jonathanj commented|
|-|-|

(In [27544]) Don't rely on the ordering of dict keys. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @jonathanj|
|-|-|

Thanks.  A few more testing issues:

  1. `parseModes` has an uncovered line, `raise IRCBadModes('Empty mode sequence: %r' % (modes,))`
  1. The last two lines of `isupport_CHANMODES` are uncovered
  1. I'm sure I must be making a mistake, but it seems `irc_MODES` has no test coverage.
  1. Likewise for `getUserModeParams`
  1. Only the positive path through `getChannelModeParams` is tested.

Since we recognized `h` before in this, it'd be good to keep recognizing it.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_19"></a>|@jonathanj commented|
|-|-|

(In [27547]) Add test coverage for empty mode sequences, malformed CHANMODES parameters and getUserModeParams. Improve test coverage for getChannelModeParams. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_20"></a>|@jonathanj commented|
|-|-|

(In [27548]) Add "h" mode to PREFIX default, maintaining functionality provided by _modeAcceptsArg. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_21"></a>|@jonathanj commented|
|-|-|

(In [27552]) Add test coverage for irc_MODE. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_22"></a>|@jonathanj set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_18):
> Thanks.  A few more testing issues:
>
>   1. `parseModes` has an uncovered line, `raise IRCBadModes('Empty mode sequence: %r' % (modes,))`
>   1. The last two lines of `isupport_CHANMODES` are uncovered
>   1. I'm sure I must be making a mistake, but it seems `irc_MODES` has no test coverage.
>   1. Likewise for `getUserModeParams`
>   1. Only the positive path through `getChannelModeParams` is tested.
>
> Since we recognized `h` before in this, it'd be good to keep recognizing it.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_23"></a>|@jonathanj commented|
|-|-|

Replying to [exarkun](#note_18):
>   3. I'm sure I must be making a mistake, but it seems `irc_MODES` has no test coverage.

Looks like removed all of the previous `irc_MODE` tests (because they were using imaginary modes that were not consistent between tests) without replacing them. I've added test coverage for `irc_MODE`.

The other points have also been addressed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to @jonathanj|
|-|-|

Thanks!

  1. `log.err(failure, message)` or `log.err(None, message)` is preferred over `log.msg(message); log.err(failure)` for newer code.
  1. `testISupport` should have a docstring, and possibly be renamed to follow the naming convention.
  1. In the `test_modeMissingDirection` docstring, I think `C{+}` et al would be better spelled with `C{'+'}`.
  1. There's a bit of extra whitespace before `parseModes`.

Tests are failing on the Python 2.3 slaves because of `sorted`.  However, we just dropped 2.3 support.  I'm investigating getting those slaves updated to a new version of Python.

Otherwise, looks good.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_25"></a>|@jonathanj commented|
|-|-|

(In [27709]) Address review commentary. refs [#3296](https://github.com/twisted/twisted/issues/3296)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set owner to @jonathanj|
|-|-|

Looks good.  Please add a news file and merge.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_28"></a>|@jonathanj commented|
|-|-|

(In [27716]) Branching to 'isupport-aware-mode-parsing-3296-2'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_29"></a>|@jonathanj set status to `closed`|
|-|-|

(In [27721]) Merge isupport-aware-mode-parsing-3296-2: Take ISUPPORT information into account when parsing mode messages.

Replace the hardcoded list of modes that accept parameters with a mechanism
that uses ISUPPORT information, as specified by the server, to determine which
modes accept parameters. The default behaviour matches that of the legacy
"_modeAcceptsArg" mechanism.

Author: jonathanj
Reviewer: exarkun
Fixes: [#3296](https://github.com/twisted/twisted/issues/3296)
